compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -266,15 +272,221 @@ static LogicalResult replaceAllStoresWithTiledVersion(
 
 namespace {
 
-/// Result of the tiled operation.
+/// Result of the tiled operation. The resulted tiled loops may have a different
+/// number of loops of the input loop range. `loops`, `tileSizesForLoops`, and
+/// `boundedSizesForLoops` have the same size.
 struct TilingResult {
   Operation *tiledOp;
-  SmallVector<scf::ForOp> loops;
+  // BitVector for the tiled loops in the loop range
   llvm::SmallBitVector tiledLoops;
+  // offsets for the input loop range
   SmallVector<OpFoldResult> tileOffsets;
+  // tile sizes for the input loop range
   SmallVector<OpFoldResult> tileSizes;
+
+  // The information about the tiled loops
+  //
+  // tiled loops.
+  SmallVector<scf::ForOp> loops;
+  // tile sizes for the resulted loops
+  SmallVector<Value> tileSizesForLoops;
```

Comment:
Coming back to this, I am still not clear what `tileSizesForLoops` and `boundedSizesForLoops` are. We shouldnt really be carrying so much state. A lot of complexity gets added to the code to just manage this state. We should recover state from IR.

Reply:
Looks like `boundedSizesForLoops` is just the loop iteration domain. You can get that from the op directly.
`tileSizesForLoops` is just the tile sizes specified. That also you can get directly. I dont see why they need to be captured here.

Reply:
I agree with you that a pattern matching (which I saw some other places too) is tricky and fragile and that's the main motivation to add additional fields to `TilingResult`. The information is recorded when the loop nest is constructed, so there is no hidden assumption about the affine min. Otherwise, I need to inspect the loop nests outside and do a tricky pattern matching to see if an op is the right one to replace.

Making a tie between the bounded size op and the structure is very intentional. I think this is a better way that a pattern matching. Patten matching is more fragile than the returned structure. We need to modify the structure when the loop nest creation code gets changed. (This is code dependency, which is preferred than the pattern dependency.) The main problem is we want an easy way to get the bounded size op. Here are some other ways to get the same result.

1. Make the op with an attribute to say that this is a bounded size op.
2. Create an explicit op for the bounded size op. This needs a separate lowering pass.

I think Option 2) is the cleanest and most formal way if we really concerned about the pattern matching. Option 1) can be a back up solution if we are hesitant to introduce an op and an additional lowering pass.

Regarding the fields in `TilingResult`, the reason I added the fields only for generated loops is to make it easy to query the information for each loop. When a different distribution method is used, a loop might not be created, for which we don't specialize the body. Generated loops and their tightly coupled information seem to be a good result to keep and return, IMO.

Reply:
I am still not following this. AFAICS this information is not required to be computed during any of the implementation of generating the loops, but are rather inputs to the computation that generates the loop. Using the inputs directly where needed is better right?

Reply:
Just to be clear, are you talking about `tileSizesForLoops`?

Reply:
I am talking about both `tileSizesForLoops` and `boundedSizesForLoops`. I have no idea what they are or why that state needs to be carried.